server/utils/parseMentions.ts

```diff
@@ -0,0 +1,38 @@
+import { Node } from "prosemirror-model";
+import { Document, Revision } from "@server/models";
+import DocumentHelper from "@server/models/helpers/DocumentHelper";
+
+/**
+ * Parse a list of mentions contained in a document or revision
+ *
+ * @param document Document or Revision
+ * @returns An array of mentions in passed document or revision
+ */
+export default function parseMentions(
+  document: Document | Revision
+): Record<string, string>[] {
+  const node = DocumentHelper.toProsemirror(document);
+  const mentions: Record<string, string>[] = [];
+  const visited: Map<string, boolean> = new Map();
+
+  function findMentions(node: Node) {
+    if (visited.get(node.attrs.id)) {
+      return;
+    }
+
+    visited.set(node.attrs.id, true);
```

Comment:
Seems fine to use a `Set` but why not just store the mentions in the values of the set rather than keeping two data structures? 😉

Reply:
The map keeps track of which nodes have already been visited during depth-first traversal, kind of like memoization to improve perf by not traversing same nodes again and again. May use `Set` to eliminate duplicates post traversal but doesn't help with efficient traversal.

Reply:
`descendants` should already be efficient, it shouldn't be visiting the same node more than once at least

Reply:
Not sure how that works but I've found for example the `doc` node(and others) being visited more than once in local testing. Either way, the `visited` checks are not completely accurate because it relies on `node.attrs.id` which is specific to `mention` nodes and not all. Hence, falling back to the `Set` approach :+1: . Wish there was a node explorer for prosemirror, would've been very helpful with debugging :disappointed: